### DISCUSSION TITLE
Any way to sacrifice precision to gain speed on earth satellite propigation?

### DISCUSSION THREAD

**Discussion by yoachim** [skyfielders/python-skyfield#756] [Q&A] [answered]
Hi all,
I'm currently looking at how things like Starlink mega constellations will impact Rubin observatory. I can construct realistic TLEs and make a list of `EarthSatellite` objects. My issue is updating the positions of ~30k satellites can get computationally expensive, and if I want to check all 2 million planned Rubin observations over 10 years for satellite streaks it will really take a while.

Is there a propagator besides `SGP4`? I'd be happy with something super simple that just assumed the orbital elements stayed constant as long as I could still compute alt,az from a location and check solar illumination.

**Comment by davidmikolas** [comment]
### Background (you likely know all this already)

1. The orbits of satellites in low Earth orbit (LEO) are not accurately predictable for more than a few days to a few weeks at most, depending on the level of accuracy you want. The major effects that make them unpredictable are atmospheric drag (yes there's air up there at 250 to 550 km) and to a much lesser extent solar photon pressure, and propulsive maneuvers produced by the satellites. They can execute significant trajectory changing burns of their propulsion system to adjust their orbital phasing with other satellites, or to move to a new orbit.
2. The SGP4 propagator is an elaborate approximation never intended to be accurate out beyond a few days to a few weeks for Earth satellites in LEO. By its very nature it requires regular supply of new, freshly prepared TLEs based on fitting much more elaborate and powerful orbital propagators that are used to generate trajectories that fit radar, optical, on-board GPS (yes these days many satellites make obtain their own GPS fixes) and other regularly measured tracking data.

### "I can construct realistic TLEs"

In that case if you appreciate that they won't be correct and actually predictive TLEs of actual satellites, but instead just want to perform a statistical analysis based on planned (or assumed, really) constellation configurations of satellites that are already at their final altitude, then you can just make a simple model of equal spaced  objects on circular orbits with inclinations and ascending nodes spaced based on stated SpaceX plans in their submissions to the FCC for licensing purposes. Note however that their plans are dynamic and change regularly, so nobody knows how many satellites they'll launch or into what orbits they'll inject them. They are learning as they go. It's all driven by business; they need to keep an eye on paying customers, Starlink must make money ASAP!

Also note that satellite brightness can be much higher during the climb phase; each launch inserts a group of about 60 Starlink satellites into a very low orbit with an altitude of about 250 to 300 km as a tumbling cluster of thin, flat, shiny objects. They wait until they drift apart a bit, then deploy their long, thin, flat, somewhat shiny solar panels, and with their newfound electrical power fire up their ion engines and make the slow climb up to their desired altitude, making sure to space themselves out along the general orbit into-which they've been injected.

This takes weeks to months. It's unpredictable because solar activity modulates the temperature of Earth's [thermosphere](https://www.nasa.gov/mission_pages/sunearth/science/atmosphere-layers2.html) and thus its density, which wanders over a factor of ten or so depending on the Sun's mood. 

One SpaceX launch completely tanked because around the time of launch there was solar flare-induced heating event that made it impossible for the satellites' ion propulsion to overcome atmospheric drag and still have enough propellant left over for a long and productive life. 

 - [SpaceX just lost 40 satellites to a geomagnetic storm. There could be worse to come](https://www.technologyreview.com/2022/02/10/1045202/spacex-just-lost-40-satellites-to-a-geomagnetic-storm-there-could-be-worse-to-come/)

From time to time some will even move over to join or resupply an adjacent orbit.

They will also be regularly reaching end of life (EOL) or malfunctioning, and so de-orbiting by lowering their altitude over weeks to a point where atmospheric drag will rapidly lower them to destruction.

There will be regular launches and regular deorbits for quite some time to come (many years).

### Conclusion

None of the information necessary for **accurate** predictions into the future is going to be available, so if you want to do a ten year impact simulation you are going to have to "roll your own" simulation based on your own assumptions.

The math necessary to calculate the apparent RA/Dec of a spot on a circular orbit from a spot on an oblate Earth is pretty simple actually, and won't require numerical integration.

But modeling the launches and the month period of altitude climbing and orbital insertion will be challenging and fun, and choosing your average albedo vs time during that phase will require some serious thinking and research.

Join or at least check out [Space Exploration Stack Exchange](https://space.stackexchange.com/questions) check the orbital-mechanics tag and the starlink tag and see what's been discussed about their launches and on-orbit behaviors.

### Skyfield is not the right environment for this project

It's not an easy project, and Skyfield is not the right approach. Skyfield shines in applications where astrometric accuracy or searches for specific events are needed, but you won't have input data good enough to make even something like SGP4 a reasonable or helpful choice, so you're slowing your calculation down a factor of 100 or 1000 by using the wrong tool.

![image](https://user-images.githubusercontent.com/17236804/176794270-cf9e512b-8803-44ea-8aff-e5f3cf9eb135.png)

From [SpaceX's 4,425 satellite constellation - what's the method to the madness?](https://space.stackexchange.com/q/19055/12102)

**Reply by yoachim** [reply]
I agree that SGP4 is probably 100-1000 times more precision than I need. But if using it means having to write 100-1000 times few lines of code, it gets tempting.

**Comment by yoachim** [comment, marked as answer]
Turns out the solution was just to propagate satellite orbits with arrays of desired times. Still have to loop over each satellite, but that's not as painful as looping over dates. Anyway, results to be published in the Astrophysical Journal Letters, preprint here: https://arxiv.org/abs/2211.15908

**Reply by davidmikolas** [reply]
_Beautiful!_ Congratulations on your ApJL and results. Fig 1. caption indicates the black dots in the middle row are those above the 20 degree elevation limit, but it looks like they avoid an additional chunk of sky where there's still some twilight or zodiacal light?

**Reply by brandon-rhodes** [reply]
_Gads_ — I had no idea the constellations were expected to get that large! That's too many satellites.

Thank you for following up with the link to the paper, and for describing how you wound up performing the calculation.

**Reply by yoachim** [reply]
> Fig 1. caption indicates the black dots in the middle row are those above the 20 degree elevation limit, but it looks like they avoid an additional chunk of sky where there's still some twilight or zodiacal light?

I think the avoidance zone your noticing is mostly caused by earth's shadow. You can see there are lots of red points above 20-degrees, those are the satellites that are up but not illuminated by the sun.  The twilight and zodiacal light wouldn't be too bad, but it's still higher airmass so the extinction and image quality wouldn't be ideal.

**Reply by davidmikolas** [reply]
Got it, thanks :-)
